### DISCUSSION TITLE
Extracting drawing to a new pdf

### DISCUSSION THREAD

**Discussion by faisalakhlaq** [pymupdf/PyMuPDF#2296] [Looking for help]
**Is your feature request related to a problem? Please describe.**
I have a PDF document containing different paths (drawings). Can I extract each existing path in a new PDF document without using Shape?
The extracted drawings should be exactly the same as in the original PDF. With all the layers (OCG), colors, embedded / attached images e.t.c. Therefore, it will be wise not to create a new Shape object (I tried to do that but it loses the colors).
 
**Describe the solution you'd like**
Something along the lines:
```
        paths = page.get_drawings()  # extract existing drawings
        for index, path in enumerate(paths):
            rect = path['rect']
            module_pdf = fitz.open()
            outpage = module_pdf.new_page(width=rect.width, height=rect.height)        
            outpage.show_pdf_page(rect, path, 0)
            module_pdf.save(f"{index}.pdf")

```
**Describe alternatives you've considered**
I tried looping over the paths and creating shapes as demonstrated in many examples but I get only the outlines of the paths.

```
shape = outpage.new_shape()

for index, path in enumerate(paths):
    if item[0] == "l":  # line
            shape.draw_line(item[1], item[2])
           ....... 
    shape.finish(
        fill=path["fill"],
        ........
    shape.commit()
            outpdf.save(f"{index}.pdf")

```

**Additional contex
Below is the screenshot of the pdf and when extracted the expected output is 2 separate files each with one drawing.

![Screenshot from 2023-03-21 12-07-58](https://user-images.githubusercontent.com/2509965/226588706-07c9eb70-0364-4f5f-a8b6-f0e45afbf2d2.png)
t**

**Comment by JorjMcKie** [comment]
Having trouble to understand what of the old PDF you **_don't_** want in your new one. Just the text?
While your request as stated doesn't sound to be ever doable, you might be better off just removing the text using redactions.

**Comment by faisalakhlaq** [comment]
Many thanks for a very quick response. Was not expecting that. Sorry if my question was not well explained.  

`Having trouble to understand what of the old PDF you don't want in your new one.Just the text?` 
There is no text, only drawings. The above image (screenshot of the pdf) has 2 drawings (If I am using the terminology correctly). I tried to extract each drawing in a new pdf. So result is 2 pdfs each containing one drawing. 

```
        paths = page.get_drawings()  # extract existing drawings
        for index, path in enumerate(paths):
            rect = path['rect']
            module_pdf = fitz.open()
            outpage = module_pdf.new_page(width=rect.width, height=rect.height)        
            outpage.show_pdf_page(rect, path, 0)
            module_pdf.save(f"{index}.pdf")

```
In the above code snippet if `outpage.show_pdf_page(rect, path, 0)` can work or something else like this exists then it solves the problem. `show_pdf_page` expects a pdf but I just want to put the path on it as I don't have any PDF.

`While your request as stated doesn't sound to be ever doable`
Good to know then I should stop exploring any further options.

**Comment by JorjMcKie** [comment]
What you **can** do is restricting the part of the source page rectangle that should be visible in the target page: `tar_page.show_pdf_page(rect, source_pdf, source_pagr_no, clip=source_rect)`.

This reduces your problem to determnining the `source_rect` - presumably the rectangle that covers one of your drawings.
But in any case, this **_will not_** reflect any OCG conditions that may exist in the source document.

**Comment by JorjMcKie** [comment]
Let me convert this issue to a "Discussions" item. Will be easier to handle there.

**Comment by JorjMcKie** [comment]
To output each drawing (vector graphic) on the pages of a source PDF to a new page of some target PDF, you can thus do the following:
```python
src = fitz.open("source.pdf")
out = fitz.open()  # output PDF having one drawing per page
for src_page in src:
    paths = src_page.get_drawings()  # vector graphics on source
    for path in paths:
        rect = path["rect"]  # rect covering the drawing
        tar_page = out.new_page(width=rect.width, height=rect.height)
        tar_page.show_pdf_page(tar_page.rect, src, src_page.number, clip=rect)
out.cose("paths.pdf")
```
